### PR TITLE
Java: Add XCLAIM command

### PR DESCRIPTION
**Pull request by acarbonetto**
https://redis.io/docs/latest/commands/xclaim/

Note: the LASTID argument is undocumented.  After multiple attempts to test it, I was unable to determine the effect of the argument.  Following in Lettuce.io footsteps and removing the option from the API.

### CODE REVIEW

File: glide-core/src/client/value_conversion.rs [hunk 1]
```diff
@@ -853,6 +853,16 @@ pub(crate) fn expected_type_for_cmd(cmd: &Cmd) -> Option<ExpectedReturnType> {
             key_type: &None,
             value_type: &None,
         }),
+        b"XCLAIM" => {
+            if cmd.position(b"JUSTID").is_some() {
+                Some(ExpectedReturnType::ArrayOfStrings)
+            } else {
+                Some(ExpectedReturnType::Map {
+                    key_type: &Some(ExpectedReturnType::BulkString),
```
Comment (Yury-Fridlyand) [inline glide-core/src/client/value_conversion.rs:861]: Maybe
```suggestion
                    key_type: &Some(ExpectedReturnType::SimpleString),
```
It is entry ID right? In timestamp-like format.

File: glide-core/src/client/value_conversion.rs [hunk 2]
```diff
@@ -853,6 +853,16 @@ pub(crate) fn expected_type_for_cmd(cmd: &Cmd) -> Option<ExpectedReturnType> {
             key_type: &None,
             value_type: &None,
         }),
+        b"XCLAIM" => {
```
Comment (Yury-Fridlyand) [inline glide-core/src/client/value_conversion.rs:856]: Add tests please

File: java/client/src/main/java/glide/api/BaseClient.java [hunk 3]
```diff
@@ -2001,6 +2003,58 @@ public CompletableFuture<Object[][]> xpending(
...
+    public CompletableFuture<String[]> xclaimJustId(
+            @NonNull String key,
+            @NonNull String group,
+            @NonNull String consumer,
+            long minIdleTime,
+            @NonNull String[] ids,
+            @NonNull StreamClaimOptions options) {
+        String[] args =
+                concatenateArrays(new String[] {key, group, consumer, Long.toString(minIdleTime)}, ids);
+        args = concatenateArrays(args, options.toArgs(true));
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/BaseClient.java:2053]: if `justid` can be the last argument, do
```suggestion
        args = concatenateArrays(args, options.toArgs(), JUSTID);
```

File: java/client/src/main/java/glide/api/BaseClient.java [hunk 4]
```diff
@@ -2001,6 +2003,58 @@ public CompletableFuture<Object[][]> xpending(
...
+    public CompletableFuture<Map<String, String[]>> xclaim(
+            @NonNull String key,
+            @NonNull String group,
+            @NonNull String consumer,
+            long minIdleTime,
+            @NonNull String[] ids,
+            @NonNull StreamClaimOptions options) {
+        String[] args =
+                concatenateArrays(new String[] {key, group, consumer, Long.toString(minIdleTime)}, ids);
+        args = concatenateArrays(args, options.toArgs(false));
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/BaseClient.java:2028]: why 2 `concatenateArrays`?

File: java/client/src/main/java/glide/api/commands/StreamBaseCommands.java [hunk 5]
```diff
@@ -672,4 +673,79 @@ CompletableFuture<Object[][]> xpending(
...
+     * Changes the ownership of a pending message.
+     *
+     * @see <a href="https://valkey.io/commands/xclaim/">valkey.io</a> for details.
+     * @param key The key of the stream.
+     * @param group The consumer group name.
+     * @param consumer The group consumer.
+     * @param minIdleTime The minimum idle time for the message to be claimed.
+     * @param ids An array of entry ids.
+     * @return An <code>array</code> of message entries with the format <code>
+     *     [[id, ["entry", "data"]], ...]</code> that are claimed by the consumer.
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/StreamBaseCommands.java:687]: A bit more js style for a map
```suggestion
     *     {entryId : [key, value], ...}</code> that are claimed by the consumer.
```

File: java/client/src/main/java/glide/api/commands/StreamBaseCommands.java [hunk 6]
```diff
@@ -672,4 +673,79 @@ CompletableFuture<Object[][]> xpending(
...
+     * @example
+     *     <pre>
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/StreamBaseCommands.java:689]: todo reminder

File: java/client/src/main/java/glide/api/commands/StreamBaseCommands.java [hunk 7]
```diff
@@ -672,4 +673,79 @@ CompletableFuture<Object[][]> xpending(
...
+     *
+     * @see <a href="https://valkey.io/commands/xclaim/">valkey.io</a> for details.
+     * @param key The key of the stream.
+     * @param group The consumer group name.
+     * @param consumer The group consumer.
+     * @param minIdleTime The minimum idle time for the message to be claimed.
+     * @param ids An array of entry ids.
+     * @param options Stream claim options {@link StreamClaimOptions}.
+     * @return An <code>array</code> of message entries with the format <code>
+     *     [[id, ["entry", "data"]], ...]</code> that are claimed by the consumer.
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/StreamBaseCommands.java:706]: same here

File: java/client/src/main/java/glide/api/commands/StreamBaseCommands.java [hunk 8]
```diff
@@ -672,4 +673,79 @@ CompletableFuture<Object[][]> xpending(
...
+     * @param consumer The group consumer.
+     * @param minIdleTime The minimum idle time for the message to be claimed.
+     * @param ids An array of entry ids.
+     * @return An <code>array</code> of message ids claimed by the consumer.
+     */
+    CompletableFuture<String[]> xclaimJustId(
+            String key, String group, String consumer, long minIdleTime, String[] ids);
+
+    /**
+     * Changes the ownership of a pending message.This command uses the JUSTID optional argument to
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/StreamBaseCommands.java:732]:
```suggestion
     * Changes the ownership of a pending message. This command uses the JUSTID optional argument to
```

File: java/client/src/main/java/glide/api/commands/StreamBaseCommands.java [hunk 9]
```diff
@@ -672,4 +673,79 @@ CompletableFuture<Object[][]> xpending(
...
+    CompletableFuture<Map<String, String[]>> xclaim(
+            String key,
+            String group,
+            String consumer,
+            long minIdleTime,
+            String[] ids,
+            StreamClaimOptions options);
+
+    /**
+     * Changes the ownership of a pending message. This command uses the JUSTID optional argument to
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/StreamBaseCommands.java:717]: It is optinal for command in Redis API, but not in this very function.
Maybe reword if you think it is needed:
> Changes ... This function makes redis command to return only message/entry ID, since it uses JUSTID command/Redis API argument.

File: java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+/**
+ * Optional arguments to {@link StreamBaseCommands#xclaim(String, String, String, long, String[],
+ * StreamClaimOptions)}
+ *
+ * @see <a href="https://redis.io/commands/xclaim/">redis.io</a>
+ */
+@Builder
+public class StreamClaimOptions {
+
+    /** Redis api string to designate IDLE time in milliseconds */
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java:18]: api -> API

File: java/client/src/main/java/glide/api/commands/StreamBaseCommands.java (hunk 9 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/commands/StreamBaseCommands.java:717]: Wrap JUSTID with `code`

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 11]
```diff
@@ -3276,6 +3278,103 @@ public T xpending(
...
+     * Changes the ownership of a pending message.
+     *
+     * @see <a href="https://valkey.io/commands/xclaim/">valkey.io</a> for details.
+     * @param key The key of the stream.
+     * @param group The consumer group name.
+     * @param consumer The group consumer.
+     * @param minIdleTime The minimum idle time for the message to be claimed.
+     * @param ids An array of entry ids.
+     * @return Command Response - An <code>array</code> of message entries with the format <code>
+     *     [[id, ["entry", "data"]], ...]</code> that are claimed by the consumer.
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:3291]: same here

File: java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+    public static final String FORCE_REDIS_API = "FORCE";
+
+    /** Redis api string to designate JUSTID */
+    public static final String JUST_ID_REDIS_API = "JUSTID";
+
+    /** Redis api string to designate LASTID */
+    public static final String LAST_ID_REDIS_API = "LASTID";
+
+    /**
+     * Set the idle time (last time it was delivered) of the message. If <code>idle</code> is not
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java:37]: > Set the idle time in ms ...

Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java:37]: > If `idle` is unset, default value of `0` is assumed ...

File: java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+     * specified, an <code>idle</code> of <code>0</code> is assumed, that is, the time count is reset
+     * because the message has now a new owner trying to process it.
+     */
+    private final Long idle; // in milliseconds
+
+    /**
+     * This is the same as idle but instead of a relative amount of milliseconds, it sets the idle
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java:44]: as `{@link #idle`

Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java:44]: Do these 2 time arguments conflict if both set?

Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java:44]: Do we need such detailed descriptions for them though?

File: java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+     * time to a specific Unix time (in milliseconds). This is useful in order to rewrite the AOF file
+     * generating <code>XCLAIM</code> commands.
+     */
+    private final Long idleUnixTime; // in unix-time milliseconds
+
+    /**
+     * Set the retry counter to the specified value. This counter is incremented every time a message
+     * is delivered again. Normally <code>XCLAIM</code> does not alter this counter, which is just
+     * served to clients when the <code>XPENDING</code> command is called: this way clients can detect
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java:53]: Consider adding clickable links to commands

File: java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+         * otherwise the IDs of non-existing messages are ignored.
+         */
+        public StreamClaimOptionsBuilder force() {
+            return isForce(true);
+        }
+    }
+
+    /**
+     * Converts options for Xclaim into a String[].
+     *
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java:83]: Tell about the arg

File: java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+     * @return String[]
+     */
+    public String[] toArgs(boolean isJustId) {
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java:86]: Do we need UT for this?

File: java/integTest/src/test/java/glide/TransactionTestUtilities.java [hunk 17]
```diff
@@ -850,6 +867,17 @@ private static Object[] streamCommands(BaseTransaction<?> transaction) {
...
+            Map.of(
+                    "0-3",
+                    new String[] {
+                        "field3", "value3"
+                    }), // xclaim(streamKey1, groupName1, consumer1, 0L, {"0-3"}, options)
+            new String[] {
+                "0-3"
+            }, // xclaimJustId(streamKey1, groupName1, consumer1, 0L, new String[] {"0-3"})
+            new String
+                    [0], // xclaimJustId(streamKey1, groupName1, consumer1, 0L, new String[] {"0-4"}, options)
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/TransactionTestUtilities.java:880]: spotless, stop this please!
Maybe you can shorten the comment to tweak that?

File: java/integTest/src/test/java/glide/cluster/ClusterTransactionTests.java [hunk 18]
```diff
@@ -85,6 +85,9 @@ public void transactions_with_group_of_commands(String testName, TransactionBuil
         Object[] expectedResult = builder.apply(transaction);
 
         Object[] results = clusterClient.exec(transaction).get();
+        if (testName.equals("Stream Commands")) {
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/cluster/ClusterTransactionTests.java:88]: Please remove before merge

File: java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java (hunk 10 above)
Comment (jduo) [inline java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java:18]: Redis -> Valkey

File: java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+    private final Long idleUnixTime; // in unix-time milliseconds
+
+    /**
+     * Set the retry counter to the specified value. This counter is incremented every time a message
+     * is delivered again. Normally <code>XCLAIM</code> does not alter this counter, which is just
+     * served to clients when the <code>XPENDING</code> command is called: this way clients can detect
+     * anomalies, like messages that are never processed for some reason after a big number of
+     * delivery attempts.
+     */
+    private final Long retryCount;
```
Comment (jduo) [inline java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java:57]: Is this a long instead of an int just due to policy? That's alot of retries.

File: java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+
+    /**
+     * Creates the pending message entry in the PEL even if certain specified IDs are not already in
+     * the PEL assigned to a different client. However, the message must exist in the stream,
+     * otherwise the IDs of non-existing messages are ignored.
+     */
+    private final boolean isForce;
+
+    /** Filter up to the <code>lastid</code> when claiming messages */
```
Comment (jduo) [inline java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java:66]: Should lastid be in all-caps? ScanOptions documentation shows code in all-caps.

File: java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+import glide.api.commands.StreamBaseCommands;
+import java.util.ArrayList;
+import java.util.List;
+import lombok.Builder;
+
+/**
+ * Optional arguments to {@link StreamBaseCommands#xclaim(String, String, String, long, String[],
+ * StreamClaimOptions)}
+ *
+ * @see <a href="https://redis.io/commands/xclaim/">redis.io</a>
```
Comment (jduo) [inline java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java:13]: This should be a valkey URL

File: java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+    /** Redis api string to designate JUSTID */
+    public static final String JUST_ID_REDIS_API = "JUSTID";
+
+    /** Redis api string to designate LASTID */
+    public static final String LAST_ID_REDIS_API = "LASTID";
+
+    /**
+     * Set the idle time (last time it was delivered) of the message. If <code>idle</code> is not
+     * specified, an <code>idle</code> of <code>0</code> is assumed, that is, the time count is reset
+     * because the message has now a new owner trying to process it.
```
Comment (aaron-congo) [inline java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java:39]: nit:
```suggestion
     * because the message now has a new owner trying to process it.
```

File: java/client/src/main/java/glide/api/commands/StreamBaseCommands.java [hunk 23]
```diff
@@ -672,4 +673,79 @@ CompletableFuture<Object[][]> xpending(
...
+    /**
+     * Changes the ownership of a pending message.
+     *
+     * @see <a href="https://valkey.io/commands/xclaim/">valkey.io</a> for details.
+     * @param key The key of the stream.
+     * @param group The consumer group name.
+     * @param consumer The group consumer.
+     * @param minIdleTime The minimum idle time for the message to be claimed.
+     * @param ids An array of entry ids.
+     * @return An <code>array</code> of message entries with the format <code>
```
Comment (aaron-congo) [inline java/client/src/main/java/glide/api/commands/StreamBaseCommands.java:686]:
```suggestion
     * @return A <code>Map</code> of message entries with the format <code>
```

File: java/client/src/main/java/glide/api/commands/StreamBaseCommands.java [hunk 24]
```diff
@@ -672,4 +673,79 @@ CompletableFuture<Object[][]> xpending(
...
+     * Changes the ownership of a pending message.
+     *
+     * @see <a href="https://valkey.io/commands/xclaim/">valkey.io</a> for details.
+     * @param key The key of the stream.
+     * @param group The consumer group name.
+     * @param consumer The group consumer.
+     * @param minIdleTime The minimum idle time for the message to be claimed.
+     * @param ids An array of entry ids.
+     * @param options Stream claim options {@link StreamClaimOptions}.
+     * @return An <code>array</code> of message entries with the format <code>
```
Comment (aaron-congo) [inline java/client/src/main/java/glide/api/commands/StreamBaseCommands.java:705]:
```suggestion
     * @return A <code>Map</code> of message entries with the format <code>
```

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 25]
```diff
@@ -3276,6 +3278,103 @@ public T xpending(
...
+    /**
+     * Changes the ownership of a pending message.
+     *
+     * @see <a href="https://valkey.io/commands/xclaim/">valkey.io</a> for details.
+     * @param key The key of the stream.
+     * @param group The consumer group name.
+     * @param consumer The group consumer.
+     * @param minIdleTime The minimum idle time for the message to be claimed.
+     * @param ids An array of entry ids.
+     * @return Command Response - An <code>array</code> of message entries with the format <code>
```
Comment (aaron-congo) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:3290]:
```suggestion
     * @return Command Response - A <code>Map</code> of message entries with the format <code>
```

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 26]
```diff
@@ -3276,6 +3278,103 @@ public T xpending(
...
+     * Changes the ownership of a pending message.
+     *
+     * @see <a href="https://valkey.io/commands/xclaim/">valkey.io</a> for details.
+     * @param key The key of the stream.
+     * @param group The consumer group name.
+     * @param consumer The group consumer.
+     * @param minIdleTime The minimum idle time for the message to be claimed.
+     * @param ids An array of entry ids.
+     * @param options Stream claim options {@link StreamClaimOptions}.
+     * @return Command Response - An <code>array</code> of message entries with the format <code>
```
Comment (aaron-congo) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:3312]:
```suggestion
     * @return Command Response - A <code>Map</code> of message entries with the format <code>
```

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 27]
```diff
@@ -4392,6 +4393,23 @@ public void xpending(BaseClient client) {
                 ArrayUtils.remove(pending_results_extended[4], 2));
         assertTrue((Long) pending_results_extended[4][2] >= 0L);
 
+        // use claim to claim stream 3 and 4 for consumer 1
```
Comment (aaron-congo) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:4396]:
```suggestion
        // use claim to claim stream 3 and 5 for consumer 1
```

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 28]
```diff
@@ -4586,6 +4603,145 @@ public void xpending_return_failures(BaseClient client) {
...
+        String streamid_1 = client.xadd(key, Map.of("field1", "value1")).get();
+        assertNotNull(streamid_1);
+        assertNotNull(client.xreadgroup(Map.of(key, ">"), groupName, consumer1).get());
+
+        // claim with invalid stream entry IDs
+        ExecutionException executionException =
+                assertThrows(
+                        ExecutionException.class,
+                        () ->
+                                client.xclaimJustId(key, groupName, consumer1, 1L, new String[] {"invalid"}).get());
```
Comment (aaron-congo) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:4637]: Do we need to test this with the other function signatures as well or is it not necessary?

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 29]
```diff
@@ -4586,6 +4603,145 @@ public void xpending_return_failures(BaseClient client) {
...
+                                                stringkey,
+                                                groupName,
+                                                consumer1,
+                                                1L,
+                                                new String[] {streamid_1},
+                                                claimOptions)
+                                        .get());
+        assertInstanceOf(RequestException.class, executionException.getCause());
+    }
+
```
Comment (aaron-congo) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:4744]: Do we need to test any of the following, or not necessary?
- non-existing group
- non-existing consumer
- negative and/or 0 value for min-idle
- negative retry count
- invalid last ID

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 30]
```diff
@@ -4586,6 +4603,145 @@ public void xpending_return_failures(BaseClient client) {
...
+                assertThrows(
+                        ExecutionException.class,
+                        () ->
+                                client
+                                        .xclaim(stringkey, groupName, consumer1, 1L, new String[] {streamid_1})
+                                        .get());
+        assertInstanceOf(RequestException.class, executionException.getCause());
+        assertTrue(executionException.getMessage().contains("NOGROUP"));
+
+        final var claimOptions = StreamClaimOptions.builder().idle(1L).build();
```
Comment (aaron-congo) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:4651]: In these tests and the transaction ones we only test the idle and force options, do we need to use all options at least once in the integ tests or is it not necessary?

File: java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java (hunk 13 above)
Comment (acarbonetto) [reply java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java]: I just copied from the ValKey documentation and didn't feel it was necessary to fiddle with it.

File: java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java (hunk 19 above)
Comment (acarbonetto) [reply java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java:54]: I would hope users wouldn't set it that high.  I can switch to a long, but we convert to a string before we send it off to Redis anyways...

File: java/integTest/src/test/java/glide/SharedCommandTests.java (hunk 28 above)
Comment (acarbonetto) [reply java/integTest/src/test/java/glide/SharedCommandTests.java:4663]: I've added option function signature with force and retries.

File: java/integTest/src/test/java/glide/SharedCommandTests.java (hunk 30 above)
Comment (acarbonetto) [reply java/integTest/src/test/java/glide/SharedCommandTests.java:4681]: not necessary... we do an idle-time test above and that should be good enough.

File: java/integTest/src/test/java/glide/SharedCommandTests.java (hunk 29 above)
Comment (acarbonetto) [reply java/integTest/src/test/java/glide/SharedCommandTests.java:4774]: 1. non-existing group is tested with a non-existent key (it's the same test). 
2. non-existing consumer creates the consumer. 
3. Turns out, negative values are accepted.  Since they are minimums, they just do the same as the defaults. 
4. negative values are... ignored :(
5. I've removed last ID, as it doesn't appear to be an external option and is not documented (not exactly sure... but I also wasn't able to get it work through various tests). 
Lettice.io also doesn't have Last ID available. 
6. I can add this test too...

File: java/integTest/src/test/java/glide/TransactionTestUtilities.java (hunk 17 above)
Comment (acarbonetto) [reply java/integTest/src/test/java/glide/TransactionTestUtilities.java]: lol

File: java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java (hunk 16 above)
Comment (acarbonetto) [reply java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java]: no

File: java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java (hunk 20 above)
Comment (acarbonetto) [reply java/client/src/main/java/glide/api/models/commands/stream/StreamClaimOptions.java]: Probably.  Although, I've removed the option as it doesn't have any documentation and doesn't seem to have any visible effect.

File: glide-core/src/protobuf/redis_request.proto [hunk 31]
```diff
@@ -238,6 +238,7 @@ enum RequestType {
     FunctionRestore = 197;
     XPending = 198;
     XGroupSetId = 199;
+    XClaim = 200;
```
Comment (jduo) [inline glide-core/src/protobuf/redis_request.proto:241]: Probably worth rebasing on main now.

**Review by jduo — APPROVED** [review]

File: java/client/src/main/java/glide/api/commands/StreamBaseCommands.java [hunk 32]
```diff
@@ -672,4 +673,79 @@ CompletableFuture<Object[][]> xpending(
...
+     * Changes the ownership of a pending message.
+     *
+     * @see <a href="https://valkey.io/commands/xclaim/">valkey.io</a> for details.
+     * @param key The key of the stream.
+     * @param group The consumer group name.
+     * @param consumer The group consumer.
+     * @param minIdleTime The minimum idle time for the message to be claimed.
+     * @param ids A array of entry ids.
+     * @return A <code>Map</code> of message entries with the format <code>
+     *     {"entryId": ["entry", "data"], ...}</code> that are claimed by the consumer.
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/StreamBaseCommands.java:687]: Oh wait it is a 2D array like we have in `xread`:
```
127.0.0.1:6379> XCLAIM mystream mygroup Alice 3600000 1719710679916-0 1719710688676-0
1) 1) "1719710679916-0"
   2) 1) "foo"
      2) "bar"
      3) "foo"
      4) "bar2"
      5) "some"
      6) "value"
2) 1) "1719710688676-0"
   2) 1) "foo"
      2) "bar2"
```
So it should have the same return type with 2D array (array of pairs)

Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/commands/StreamBaseCommands.java:687]: Probably you can copy doc from `xread` then and/or update both
